### ISSUE TITLE
Possible issue with ID3 titles containing a forward slash

### ISSUE THREAD

**Issue by MrNope** [Borewit/music-metadata#52]
ID3v2.3 title tags containing forward slash characters appear to be parsed as multiple separate title tags, breaking on the slash. Is this expected behaviour / part of the ID3 spec, or a bug?

I have not noticed this behaviour in any other player / tag editor / parser when opening the same file.

Example:
A song with the title "This/That" returns the following two ID3v2.3 native tags:
```
[{
  id: "TIT2",
  value: "This"
}, {
  id: "TIT2",
  value: "That
}]
```

Thank you!

**Comment by Borewit** [comment]
So if the ID3 version is  2.3 (`major === 3)` it will treat the `/` as a value separator:

https://github.com/Borewit/music-metadata/blob/26b5895dc0c7144a6589d7d13433ffb1e837a267/src/id3v2/FrameParser.ts#L186-L189

Why not 2.4? Well they introduced an official separator. (the` x00`)

> Is this expected behaviour

Yes... but in the title.. maybe not

> part of the ID3 spec

I don't think this it is part of the ID3v2.3 specification, but it is more like a convention, I have seen it somewhere written....

> or a bug?

> I have not noticed this behaviour in any other player / tag editor / parser when opening the same file.
[Picard](https://picard.musicbrainz.org/) is using it to separate between multiple artists.

And I do want to be able to decode Picards data. The way the common data is mapped, is mostly based on Picard, I think they have done an excellent job.

What I can do is to test if a field value is a singleton (a field which is expected only to occur once in a song; like a title, but not an artist). If the field is a singleton I don't slit it on the slash?
What do you think?

**Comment by MrNope** [comment]
Here's my **opinion**, so feel free to disregard:

- using the slash character as a separator is probably bad in general, as it's ambiguous; In my own music collection I've found songs with Titles, Artists, and Album Titles including slashes as an actual part of the name.
- some tagging utilities use null characters instead for this (e.g. mp3Tag), which I think is fine, as those characters would never legitimately appear in a title, etc. it appears you are splitting on those as well, which I like.
- lastly: I don't like splitting a value on the slash because it's destructive - the application reading the parsed tags can't tell whether the tags "This" and "That" were originally "This / That" or if they were actually two separate tags to begin with. However - it would be very easy for an application to do that splitting itself, instead of this library making those assumptions

My suggested solution is therefore:
- don't split values on arbitrary characters that could legitimately appear in the tag (or at least offer an option to configure this)
- applications reading the output are free to then split the data however they want to, with no assumptions yet made

What are your thoughts on this? I'm happy to have more discussion.

Thank you for your responsiveness too!

**Comment by Borewit** [comment]
According to the IDv2.3 spec:

> [TCOM
> The 'Composer(s)' frame is intended for the name of the composer(s). **They are seperated with the "/" character**.](http://id3.org/id3v2.3.0#line-290)

Same for: 

- TEXT: The 'Lyricist(s)/Text writer(s)
- TOLY: 'Original lyricist(s)/text writer(s)'
- TOPE: Original artist(s)/performer(s)
- TPE1: Lead artist(s)/Lead performer(s)/Soloist(s)/Performing group

In other words, fields which are likely to contain one or more artistsof any kind, artists are separated by a slash forward.
But it does not make sense to split the TIT2, the _Title/Songname/Content description_, I will consider that as a bug.

> some tagging utilities use null characters instead

A null character is used as separator in ID3v2.4.
If multiple artists are encoded that way in a ID3v2.3 tag, it is clearly wrong.

So how should we encode "AC/DC" as an artist in ID3v2.2? 

Good question.

Why on earth do you want to have AC/DC in your music collection?

Related thread: [hydrogenaud.io:  Topic: How to get forward-slash / in tags to show correctly?](https://hydrogenaud.io/index.php/topic,62897.0.html)

I propose to fix this bug by splitting the ID3v2.3 TCOM, TEXT, TOLY, TOPE & TPE1 tag, and only these tags, based on the "/" character, according to the [http://id3.org/id3v2.3.0](ID3v2.3 standard).

Does that make sense?

**Comment by Borewit** [comment]
Most tools indeed completely ignore the '/', even Picard.

Life is confusing sometimes.

**Comment by MrNope** [comment]
I figure that if the ID3v2.3 spec says that those five fields should be interpreted in such a way (please excuse my ignorance of the ID3 spec in general), that's fair enough; I wouldn't expect those fields to reasonably have slashes as part of the name anyway, since they essentially refer to people's real names.

If you were to remove splitting on slashes from all other fields, I think that would be a good solution.
I guess the important thing to remember is that **the application using this library's output can still split them afterwards if it wishes**, so the ability to have that functionality is not lost.

I would be happy with your proposed solution, thanks again for your responsiveness and also for your positive engagement with my concerns! 👍

**Comment by Borewit** [comment]
No problem, great to receive some constructive feedback.

> since they essentially refer to people's real names

Artists are referred by their artist name (e.g. "AC/DC" in TPE1), which may or may not be the same as their real name.

If most implementations ignore this separator, it may be useful to offer an option to disable the artist splitting in ID3v2.3.

For the time being I will apply fix strictly according to spec on _those five fields_ .

**Comment by Borewit** [comment]
Implemented in [v0.9.5](https://github.com/Borewit/music-metadata/releases/tag/v0.9.5)

Although, this remains an arguable issue:

https://github.com/Borewit/music-metadata/blob/a7564461645cba81d33d0fd46ad4ba4b40cb0203/test/test-id3v2.3.ts#L153-L155
